### PR TITLE
Technical design of security, authorization

### PR DESCRIPTION
**Pull request by techsmyth**
Updated main Technical Design document
Updated usage of visuals
New document specific to Authorization based on Credentials
Diagrams updated (more to be done later)
...

### CODE REVIEW

File: docs/credential-based-authorization.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+The high level design for the Authorization Framework is shown below. 
+<p align="center">
+<img src="images/security-authorization-framework.png" alt="Credential Based Authorization" width="600" />
+</p>
+
+It is important to note that while the set of Credentials held by an Agent is expected to be dynamic and evolve with platform usage, the Authorization Policies are expected to not change often. 
+
+## **Why use Credentials? Are Access Control Lists / Groups not enough?**
+To operate the platform in a decentralized manner, the Authorization framewokr used cannot rely on centralized access control. Given the platforms focus on SSI, the choice was made to move to using Credentials for the authorization of all actions on the platform.
```
Comment (valentinyanakiev) [inline docs/credential-based-authorization.md]: framework --> typo

File: docs/credential-based-authorization.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+* **Access Control Lists**: the person goes to the conference entrance, and identifies themselves. Their identity is checked against a list at the door by the representative. If the *identity* is on the list the user is granted access.
+* **Credentials**: the person goes to the conference entrance, and presents a *credential* tied to their identity. If that credential is valid, and the conference rules state that a person holding a valid instance of that credential can access the conference, then the person is granted access.
+
+Note that the latter is inherently more scalable and decentralized as there is not a centralized list that is used at the entrance to the conference. 
+
+The medium term goal is to move to using full SSI with Verified Credentials. However as an intermediate step the platform uses a lightweight implementation of non-verified Credentials (i.e. not cryptographically secure and managed by the platform). This works as the platform is still centralized but it is important to switch already to Credential based authorization now so that the path to using decentralized authorisation is clear.
```
Comment (valentinyanakiev) [inline docs/credential-based-authorization.md]: usage of authorization and authorisation is inconsistent - stick with one

File: docs/credential-based-authorization.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+Note that the latter is inherently more scalable and decentralized as there is not a centralized list that is used at the entrance to the conference. 
+
+The medium term goal is to move to using full SSI with Verified Credentials. However as an intermediate step the platform uses a lightweight implementation of non-verified Credentials (i.e. not cryptographically secure and managed by the platform). This works as the platform is still centralized but it is important to switch already to Credential based authorization now so that the path to using decentralized authorisation is clear. 
+
+## **Terminology**
+The terminology used within Alkemio for Credential Based Authorization has borrowed heavily from initiatives such as [Cability Based Authorization System](https://essif-lab.eu/capability-based-authorization-system-by-jolocom/). This is to both learn from the experience within the wider SSI field, as well as to make the subsequent move to using Verified Credentials easier. 
+
+Terms:
+* **Credential**: a combination of a **Type** and **Resource** that can be assigned to an Agent.  
+* **Resource**: the Entity __to__ which the credential is bound e.g. Challenge
```
Comment (valentinyanakiev) [inline docs/credential-based-authorization.md:27]: Side note for discussion: I was thinking about ARN and how easy it is to understand, structurally, what the identifier is about. Maybe we can use something similar? Also adding wildcards in the ARN will make it possible to use credentials tied to entitties (or groups of entities) consistently, and not have two paths (without and with an identifier).

File: docs/technical-design.md [hunk 4]
```diff
@@ -142,55 +141,51 @@ The following locations are identified for the storage of data associated with A
...
+This section describes how users interact with the platform in a secure manner. 
 
-## Users
-Each User of the platform has a their own profile, which then governs their interactions with the platform. 
+The key design goals driving the setup below are:
+* **Security**: security is critical for a platform that is facilitating the collabration between multiple stakeholders: the trust is needed for interactions.
+* **Decentralization ready**: the interaction pattern has to be able to support operating in a decentralized setup. By decentralized in this case we focus on having entities such as Challenges, Opportunities, Users, Organisations being able to interaction *without* being on the same server. The initial implementation does have all entities on the same server, but the interaction pattern has to support being decentralized. 
+* **Extendable**: there are many ways that entities (e.g. users) can authenticate themselves, so the design needs to be flexible in terms of how entities are authenticated.
```
Comment (valentinyanakiev) [inline docs/technical-design.md:151]: Extensible

File: docs/technical-design.md [hunk 5]
```diff
@@ -142,55 +141,51 @@ The following locations are identified for the storage of data associated with A
...
-
-<p>
-<img src="images/security-authentication-providers.png" alt="Security: Authentication Providers" width="600" />
-</p>
+The Authentication of users is handled by Authentication Providers, a pluggable mechanism whereby multiple types of Authentication can be supported by the platform. After succesful Authentication, the platform retrieves the Agent for the User and that Agent is then used to carry out actions on behalf of the User. 
 
-The currently supported Authentication Providers are:
-* Azure Active Directory (AAD) - a production, enterprise class Identity Provider from Microsoft.
-* Demo Authentication Provider - a non-production authentication provider to show the capabilities of the platform in a lightweight way.
+The platform currently supports [Ory Kratos and Ory Oathkeeper](https://ory.sh) for Authentication. The Ory platform is an open source, enterprise class Identity Provider. Note that the platform supports OAuth2 so that it is straightforward to integrate external authentications, e.g. Github, Google etc, into the platform.
```
Comment (valentinyanakiev) [inline docs/technical-design.md]: In the particular case it's correct to say it uses OIDC and not OAuth.

**Review by valentinyanakiev — CHANGES_REQUESTED** [review]
Comments inline. Also:

1. In contributing.md:
- Going through the different pages, the alkemio repo is still referred to as 'Coordination' in contributing.md.
-  it is mentioned that Infrastructure repo is the main repo for managing environments, deployments and builds. That's infrastructure-operations now.
- It will be nice to add Nikola and Svetoslav to the 'Thank you' notes.
2. In conceptual-design.md
- (not urgent fix required, but makes an impression) - images have 3 different color schemes, doesn't make good impression.
3. In technical-design.md
- I really don't get what you mean by the 'Authentication Zone' in 'Authentication Providers' section. It seems like Kratos is outside of it and actually Kratos is the Identity Management solution and by default, if no 3rd party IDP is used, it _is_ the Authentication Zone. If you meant Authorization Zone on the diagram, that would make more sense. Also, all the images have transparent background but this.
4. In development-guidelines.md
- Shared Environments need to be updated, it's not up to date.
- Builds - same.
5. In security.md 
- Whole md is not up-to-date imo

I understand not all are in scope, but went through the whole branch and checked what's broken and not up-to-date. Fix the technical design issues and the rest we can capture on the backlog and pick up when appropriate.

File: docs/credential-based-authorization.md (hunk 1 above)
Comment (techsmyth) [reply docs/credential-based-authorization.md]: addressed in 35e6d9cd0dc2687b5eb1d60b850d806a43162dab

File: docs/credential-based-authorization.md (hunk 2 above)
Comment (techsmyth) [reply docs/credential-based-authorization.md]: addressed in 35e6d9cd0dc2687b5eb1d60b850d806a43162dab

File: docs/technical-design.md (hunk 4 above)
Comment (techsmyth) [reply docs/technical-design.md:151]: disagree, I think extendable reads fine :)

File: docs/credential-based-authorization.md (hunk 3 above)
Comment (techsmyth) [reply docs/credential-based-authorization.md:27]: Very interesting this, needs a conversation

File: docs/technical-design.md (hunk 5 above)
Comment (techsmyth) [reply docs/technical-design.md]: Fixed in 35e6d9cd0dc2687b5eb1d60b850d806a43162dab

**Comment by techsmyth** [comment]
Thanks for the comments, have addressed most of them - and put a couple over into the Epic for later work

**Review by valentinyanakiev — APPROVED** [review]

**Pull request merged**